src/job/batch_signing.rs

```diff
@@ -123,42 +118,29 @@ pub async fn execute(
         signing_sessions.update_sessions(&mut tx, &sessions).await?;
         tx.commit().await?;
     }
-
-    tracing::Span::current().record("stalled", stalled);
-    if let Some(err) = last_err {
-        return Err(err.into());
-    } else if stalled {
-        return Ok((data, false));
-    }
-
     let mut sessions = sessions.into_values();
-    let mut first_psbt = sessions
-        .next()
-        .and_then(|s| s.signed_psbt().cloned())
-        .ok_or(JobError::PsbtMissingInSigningSessions)?;
-    for s in sessions {
-        first_psbt.combine(
-            s.signed_psbt()
-                .ok_or(JobError::PsbtMissingInSigningSessions)?
-                .clone(),
-        )?;
-    }
-
-    if current_keychain.is_none() {
-        let batch = batches.find_by_id(data.account_id, data.batch_id).await?;
-        let span = tracing::Span::current();
-        span.record("txid", &tracing::field::display(batch.bitcoin_tx_id));
-        let wallet_id = batch.wallet_summaries.into_keys().next().unwrap();
-        let wallet = wallets.find_by_id(wallet_id).await?;
-        current_keychain = Some(wallet.current_keychain_wallet(&pool));
+    if let Some(mut first_signed_psbt) = sessions.find_map(|s| s.signed_psbt().cloned()) {
+        sessions.for_each(|s| {
+            if let Some(psbt) = s.signed_psbt() {
+                let _ = first_signed_psbt.combine(psbt.clone());
+            }
+        });
+        if current_keychain.is_none() {
+            let batch = batches.find_by_id(data.account_id, data.batch_id).await?;
+            let span = tracing::Span::current();
+            span.record("txid", &tracing::field::display(batch.bitcoin_tx_id));
+            let wallet_id = batch.wallet_summaries.into_keys().next().unwrap();
+            let wallet = wallets.find_by_id(wallet_id).await?;
+            current_keychain = Some(wallet.current_keychain_wallet(&pool));
+        }
+        let tx = current_keychain
+            .unwrap()
+            .finalize_psbt(first_signed_psbt)
+            .await?
```

Comment:
if the finalisation fails (because we don't have enough signatures for example) - we must return the `last_err` if there is one.